File: src/xt/db/base_flatten.clj
(ns xt.db.base-flatten
  (:require [std.lang :as l])
  (:refer-clojure :exclude [flatten]))

(l/script :xtalk
  {:require [[xt.lang.common-spec :as xt]
             [xt.lang.common-data :as xtd]
             [xt.db.base-schema :as sch]]})

(defn.xt flatten-get-links
  "flatten links"
  {:added "4.0"}
  [obj]
  (var link-fn
       (fn [e]
         (if (xt/x:is-string? e)
           (return [e true])
           (if (xt/x:nil? e)
             (xt/x:err (xt/x:cat "Invalid link - " (xt/x:json-encode obj)))
             (return [(xt/x:get-key e "id") true])))))
  (return
   (xtd/obj-keep obj
               (fn:> [v]
                     (:? (xt/x:is-array? v)
                         (-> (xt/x:arr-map v link-fn)
                             (xt/x:obj-from-pairs))
                         nil)))))

(defn.xt flatten-merge
  "flatten data"
  {:added "4.0"}
  [table-map data-obj ref-links rev-links]
  (var #{id} := data-obj)
  (var rec := (xt/x:get-key table-map id))
  (when (not (xt/x:is-object? rec))
    (:= rec {:id id
             :data {}
             :ref-links {}
             :rev-links {}})
    (xt/x:set-key table-map id rec))
  (xt/x:obj-assign (xt/x:get-key rec "data") data-obj)
  (xtd/swap-key rec "ref_links" xtd/obj-assign-with [ref-links xt/x:obj-assign])
  (xtd/swap-key rec "rev_links" xtd/obj-assign-with [rev-links xt/x:obj-assign])
  (return table-map))

(defn.xt flatten-node
  "flatten node"
  {:added "4.0"}
  [schema table-name data parent acc]
  (:= data  (xt/x:obj-assign data (xtd/clone-nested parent)))
  (var table-map (xt/x:get-key acc table-name))
  (when (not (xt/x:is-object? table-map))
    (:= table-map {})
    (xt/x:set-key acc table-name table-map))
  (var data-obj     (xtd/obj-pick data (sch/data-keys schema table-name)))
  (var obj-fn       (fn:> [v] (:? (xt/x:is-object? v) [v] v)))
  (var rev-obj      (-> (xtd/obj-pick data (sch/rev-keys schema table-name))
                        (xtd/obj-keep obj-fn)))
  (var rev-links    (-/flatten-get-links rev-obj))
  (var ref-obj      (-> (xtd/obj-pick data (sch/ref-keys schema table-name))
                        (xtd/obj-keep obj-fn)))
  (var ref-links    (-/flatten-get-links ref-obj))
  (var ref-id-map   (sch/ref-id-keys schema table-name))
  (var ref-id-links {})
  (xt/for:object [[id-k k] ref-id-map]
            (if (xt/x:is-string? (xt/x:get-key data id-k))
              (xt/x:set-key ref-id-links k {(xt/x:get-key data id-k) true})))
  (-/flatten-merge table-map
                   data-obj
                   (xtd/obj-assign-with ref-links ref-id-links xt/x:obj-assign)
                   rev-links)
  (return {:table-map table-map
           :data-obj data-obj
           :ref-obj ref-obj
           :rev-obj rev-obj}))

(defn.xt flatten-linked
  "flatten linked for schema"
  {:added "4.0"}
  [schema table-name link-obj link-id acc flatten-fn]
  (var link-fn
       (fn [e]
         (var ref (xtd/get-in schema [table-name e "ref"]))
         (return [(xt/x:get-key ref "ns")
                  (xt/x:get-key ref "rval")])))
  (xt/for:object [[e v] link-obj]
    (when (xt/x:is-array? v)
      (var [link-key link-path] (link-fn e))
      (xt/for:array [e (xt/x:arr-filter v xt/x:is-object?)]
        (flatten-fn schema link-key
                    e 
                    {link-path [link-id]}
                    acc))))
  (return acc))

(defn.xt flatten-obj
  "flatten data for schema"
  {:added "4.0"}
  [schema table-name obj parent acc]
  (var flattened  (-/flatten-node schema table-name obj parent acc))
  (var #{table-map
         data-obj
         ref-obj
         rev-obj} flattened)
  (var link-id (xt/x:get-key data-obj "id"))
  (-/flatten-linked schema table-name rev-obj link-id acc -/flatten-obj)
  (-/flatten-linked schema table-name ref-obj link-id acc -/flatten-obj)
  (return acc))

(defn.xt flatten
  "flattens data schema"
  {:added "4.0"}
  [schema table-name data parent]
  (var input (:? (xt/x:is-array? data)
                 data
                 (:? (xt/x:is-object? data)
                     data
                     [])))
  (var parent-obj (:? (xt/x:is-object? parent)
                      parent
                      {}))
  (var acc {})
  (if (xt/x:is-array? input)
    (xt/for:array [subdata input]
      (when (xt/x:not-nil? subdata)
        (-/flatten-obj schema table-name subdata parent-obj acc)))
    (-/flatten-obj schema table-name input parent-obj acc))
  (return acc))

(defn.xt flatten-bulk
  "flattens bulk data"
  {:added "4.0"}
  [schema m]
  (var acc {})
  (if (xt/x:is-array? m)
    (xt/for:array [e m]
      (var [table-name arr] e)
      (var items (:? (xt/x:is-array? arr) arr [arr]))
      (xt/for:array [obj items]
        (when (xt/x:not-nil? obj)
          (-/flatten-obj schema table-name obj {} acc))))
    (xt/for:object [[table-name arr] m]
      (var items (:? (xt/x:is-array? arr) arr [arr]))
      (xt/for:array [obj items]
        (when (xt/x:not-nil? obj)
          (-/flatten-obj schema table-name obj {} acc)))))
  (return acc))

File: packages-gen/dart/libs/xt-db/lib/text/base-flatten.dart
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'package:xtalk_db/text/base-schema.dart' as sch;
import 'dart:convert';



flatten_get_links(obj) {
  var link_fn = (e) {
    if("String" == (e.runtimeType).toString()){
      return <dynamic>[e,true];
    }
    else{
      if(null == e){
        throw "Invalid link - " + jsonEncode(obj);
      }
      else{
        return <dynamic>[e["id"],true];
      }
    }
  };
  return xtd.obj_keep(obj,(v) {
    return ((v.runtimeType).toString().startsWith("List") || (v.runtimeType).toString().startsWith("_GrowableList")) ? xtd.obj_from_pairs(xtd.arr_map(v,link_fn)) : null;
  });
}

flatten_merge(table_map, data_obj, ref_links, rev_links) {
  var id = data_obj["id"];
  var rec = table_map[id];
  if(!(("Map" == (rec.runtimeType).toString()) || (rec.runtimeType).toString().startsWith("_Map") || (rec.runtimeType).toString().startsWith("LinkedMap"))){
    rec = <dynamic, dynamic>{
      "id":id,
      "data":<dynamic, dynamic>{},
      "ref_links":<dynamic, dynamic>{},
      "rev_links":<dynamic, dynamic>{}
    };
    table_map[id] = rec;
  }
  xtd.swap_key(rec,"data",xtd.obj_assign,<dynamic>[data_obj]);
  xtd.swap_key(rec,"ref_links",xtd.obj_assign_with,<dynamic>[
    ref_links,
    (obj, other) {
      return xtd.obj_assign(obj,other);
    }
  ]);
  xtd.swap_key(rec,"rev_links",xtd.obj_assign_with,<dynamic>[
    rev_links,
    (obj, other) {
      return xtd.obj_assign(obj,other);
    }
  ]);
  return table_map;
}

flatten_node(schema, table_name, data, parent, acc) {
  data = xtd.obj_assign(data,xtd.clone_nested(parent));
  var table_map = acc[table_name];
  if(!(("Map" == (table_map.runtimeType).toString()) || (table_map.runtimeType).toString().startsWith("_Map") || (table_map.runtimeType).toString().startsWith("LinkedMap"))){
    table_map = <dynamic, dynamic>{};
    acc[table_name] = table_map;
  }
  var data_obj = xtd.obj_pick(data,sch.data_keys(schema,table_name));
  var obj_fn = (v) {
    return (("Map" == (v.runtimeType).toString()) || (v.runtimeType).toString().startsWith("_Map") || (v.runtimeType).toString().startsWith("LinkedMap")) ? <dynamic>[v] : v;
  };
  var rev_obj = xtd.obj_keep(xtd.obj_pick(data,sch.rev_keys(schema,table_name)),obj_fn);
  var rev_links = flatten_get_links(rev_obj);
  var ref_obj = xtd.obj_keep(xtd.obj_pick(data,sch.ref_keys(schema,table_name)),obj_fn);
  var ref_links = flatten_get_links(ref_obj);
  var ref_id_map = sch.ref_id_keys(schema,table_name);
  var ref_id_links = <dynamic, dynamic>{};
  for(var entry_52249 in ref_id_map.entries){
    var id_k = entry_52249.key;
    var k = entry_52249.value;
    if("String" == (data[id_k].runtimeType).toString()){
      ref_id_links[k] = <dynamic, dynamic>{data[id_k]:true};
    }
  };
  flatten_merge(table_map,data_obj,xtd.obj_assign_with(ref_links,ref_id_links,(obj, other) {
    return xtd.obj_assign(obj,other);
  }),rev_links);
  return <dynamic, dynamic>{
    "table_map":table_map,
    "data_obj":data_obj,
    "ref_obj":ref_obj,
    "rev_obj":rev_obj
  };
}

flatten_linked(schema, table_name, link_obj, link_id, acc, flatten_fn) {
  var link_fn = (e) {
    var ref = xtd.get_in(schema,<dynamic>[table_name,e,"ref"]);
    return <dynamic>[ref["ns"],ref["rval"]];
  };
  for(var entry_52250 in link_obj.entries){
    var e = entry_52250.key;
    var v = entry_52250.value;
    if((v.runtimeType).toString().startsWith("List") || (v.runtimeType).toString().startsWith("_GrowableList")){
      var value_52251 = Function.apply((link_fn as Function),<dynamic>[e]);
      var link_key = value_52251[0];
      var link_path = value_52251[1];
      var arr_52252 = xtd.arr_filter(v,(value) {
        return ("Map" == (value.runtimeType).toString()) || (value.runtimeType).toString().startsWith("_Map") || (value.runtimeType).toString().startsWith("LinkedMap");
      });
      for(var i52253 = 0; i52253 < arr_52252.length; ++i52253){
        var e = arr_52252[i52253];
        Function.apply((flatten_fn as Function),<dynamic>[
          schema,
          link_key,
          e,
          <dynamic, dynamic>{link_path:<dynamic>[link_id]},
          acc
        ]);
      };
    }
  };
  return acc;
}

flatten_obj(schema, table_name, obj, parent, acc) {
  var flattened = flatten_node(schema,table_name,obj,parent,acc);
  var data_obj = flattened["data_obj"];
  var ref_obj = flattened["ref_obj"];
  var rev_obj = flattened["rev_obj"];
  var table_map = flattened["table_map"];
  var link_id = data_obj["id"];
  flatten_linked(schema,table_name,rev_obj,link_id,acc,flatten_obj);
  flatten_linked(schema,table_name,ref_obj,link_id,acc,flatten_obj);
  return acc;
}

flatten(schema, table_name, data, parent) {
  var input = ((data.runtimeType).toString().startsWith("List") || (data.runtimeType).toString().startsWith("_GrowableList")) ? data : ((("Map" == (data.runtimeType).toString()) || (data.runtimeType).toString().startsWith("_Map") || (data.runtimeType).toString().startsWith("LinkedMap")) ? data : <dynamic>[]);
  var parent_obj = (("Map" == (parent.runtimeType).toString()) || (parent.runtimeType).toString().startsWith("_Map") || (parent.runtimeType).toString().startsWith("LinkedMap")) ? parent : <dynamic, dynamic>{};
  var acc = <dynamic, dynamic>{};
  if((input.runtimeType).toString().startsWith("List") || (input.runtimeType).toString().startsWith("_GrowableList")){
    var arr_52274 = input;
    for(var i52275 = 0; i52275 < arr_52274.length; ++i52275){
      var subdata = arr_52274[i52275];
      if(null != subdata){
        flatten_obj(schema,table_name,subdata,parent_obj,acc);
      }
    };
  }
  else{
    flatten_obj(schema,table_name,input,parent_obj,acc);
  }
  return acc;
}

flatten_bulk(schema, m) {
  var acc = <dynamic, dynamic>{};
  if((m.runtimeType).toString().startsWith("List") || (m.runtimeType).toString().startsWith("_GrowableList")){
    var arr_52296 = m;
    for(var i52297 = 0; i52297 < arr_52296.length; ++i52297){
      var e = arr_52296[i52297];
      var value_52318 = e;
      var table_name = value_52318[0];
      var arr = value_52318[1];
      var items = ((arr.runtimeType).toString().startsWith("List") || (arr.runtimeType).toString().startsWith("_GrowableList")) ? arr : <dynamic>[arr];
      var arr_52319 = items;
      for(var i52320 = 0; i52320 < arr_52319.length; ++i52320){
        var obj = arr_52319[i52320];
        if(null != obj){
          flatten_obj(schema,table_name,obj,<dynamic, dynamic>{},acc);
        }
      };
    };
  }
  else{
    for(var entry_52341 in m.entries){
      var table_name = entry_52341.key;
      var arr = entry_52341.value;
      var items = ((arr.runtimeType).toString().startsWith("List") || (arr.runtimeType).toString().startsWith("_GrowableList")) ? arr : <dynamic>[arr];
      var arr_52342 = items;
      for(var i52343 = 0; i52343 < arr_52342.length; ++i52343){
        var obj = arr_52342[i52343];
        if(null != obj){
          flatten_obj(schema,table_name,obj,<dynamic, dynamic>{},acc);
        }
      };
    };
  }
  return acc;
}

flatten_bulk_ids(schema, lookup, m) {
  var flat = flatten_bulk(schema,m);
  return xtd.arr_keep(sch.table_order(lookup),(table_name) {
    return flat.containsKey(table_name) ? <dynamic>[
      table_name,
      xtd.arr_sort(List<dynamic>.from(( flat[table_name] ).keys),(id) {
          return id;
        },(x, y) {
          return (("String" == (x.runtimeType).toString()) || ("String" == (y.runtimeType).toString())) ? ((x).toString().compareTo((y).toString()) < 0) : (x < y);
        })
    ] : null;
  });
}